DFE-Digital/sts-plan-technology-for-your-school
# Feat: use redis for caching

## Overview

Addresses ticket [#234426](https://dfe-ssp.visualstudio.com/s190-schools-technology-services/_workitems/edit/234426)

Modifies the codebase to start using the Redis cache that support was added for in a previous PR, but was not yet in use

This is intended to be the smallest amount of change possible to get it working

**This won't be merged on its own but when the cleanup piece is ready as well**

## Changes

### Major
- Any calls to the database for CMS content has been removed in favour of retrieving the content from Contentful
- Created a new `ICmsCache` inheriting from `IDistributedCache` to include methods specific to Contentful around invalidating the cache upon changes to Content
- Removed or amended tests that no longer passed due to this change
- There is still a lot of refactoring to do here to remove now redundant code, but there will be another PR for this to reduce strain on review

### Non-functional changes 
- Made some changes to existing documentation on Contentful->DB processes and changed this to explain how we use the Redis cache to store items and now use the webhook to invalidate them
- More to do here, which will come in the follow on work to cleanup

## How to review the PR

1. Follow the instructions in the Redis Readme to connect to our Azure redis or your own
2. Run plan-tech, and test that the logs show it retrieves items from the cache once they've been retrieved once
3. Try loading the start page (for example), then updating some content within the page in Contentful, the cache should invalidate the right items and show the correct updated content when the page is refreshed.

## Release requirements

Running the terraform to create the cache on test/staging/prod and setting the connection string in the keyvault for each environment

## Additional notes
## Checklist
